File: backend/app/schemas/multitrack_project.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum

class TrackType(str, Enum):
    dialogue = "dialogue"
    environment = "environment" 
    background = "background"

class ProjectInfo(BaseModel):
    id: str = Field(..., description="项目唯一标识")
    title: str = Field(..., description="项目标题")
    description: Optional[str] = Field(None, description="项目描述")
    author: Optional[str] = Field(None, description="项目作者")
    totalDuration: float = Field(0.0, description="总时长（秒）")
    sampleRate: int = Field(44100, description="采样率")
    channels: int = Field(2, description="声道数")
    bitDepth: int = Field(16, description="位深度")
    exportFormat: str = Field("wav", description="导出格式")
    createdAt: Optional[datetime] = Field(None, description="创建时间")
    version: str = Field("1.0", description="格式版本")

class Character(BaseModel):
    id: str = Field(..., description="角色唯一标识")
    name: str = Field(..., description="角色名称")
    voice: str = Field(..., description="语音模型标识")
    color: str = Field(..., description="角色颜色")
    avatar: Optional[str] = Field(None, description="角色头像路径")

class AudioClip(BaseModel):
    id: str = Field(..., description="片段唯一标识")
    name: str = Field(..., description="片段名称")
    filePath: str = Field(..., description="音频文件路径")
    startTime: float = Field(..., description="开始时间（秒）")
    duration: float = Field(..., description="持续时间（秒）")
    volume: float = Field(1.0, description="片段音量")
    fadeIn: float = Field(0.0, description="淡入时长（秒）")
    fadeOut: float = Field(0.0, description="淡出时长（秒）")
    playbackRate: float = Field(1.0, description="播放速度")
    loop: Optional[bool] = Field(False, description="是否循环播放")
    character: Optional[Character] = Field(None, description="角色信息（对话轨专用）")
    text: Optional[str] = Field(None, description="原始文本内容")
    metadata: Optional[Dict[str, Any]] = Field(None, description="扩展元数据")

class Track(BaseModel):
    id: str = Field(..., description="轨道唯一标识")
    name: str = Field(..., description="轨道名称")
    type: TrackType = Field(..., description="轨道类型")
    volume: float = Field(1.0, description="轨道整体音量")
    muted: bool = Field(False, description="是否静音")
    solo: bool = Field(False, description="是否独奏")
    color: str = Field(..., description="轨道颜色")
    order: int = Field(..., description="轨道排序")
    clips: List[AudioClip] = Field(default_factory=list, description="音频片段列表")

class Marker(BaseModel):
    id: str = Field(..., description="标记唯一标识")
    name: str = Field(..., description="标记名称")
    time: float = Field(..., description="时间点（秒）")
    type: str = Field("marker", description="标记类型")
    color: str = Field(..., description="标记颜色")

class MultitrackProject(BaseModel):
    project: ProjectInfo = Field(..., description="项目信息")
    tracks: List[Track] = Field(default_factory=list, description="音轨列表")
    markers: Optional[List[Marker]] = Field(default_factory=list, description="时间标记")

# API响应模型
class MultitrackProjectResponse(BaseModel):
    success: bool = Field(True, description="操作是否成功")
    data: Optional[MultitrackProject] = Field(None, description="项目数据")
    message: Optional[str] = Field(None, description="响应消息")

class MultitrackProjectRequest(BaseModel):
    project: MultitrackProject = Field(..., description="多音轨项目数据")

# 转换工具的数据模型
class DialogueData(BaseModel):
    """原始角色对话数据格式"""
    characters: List[Dict[str, Any]]
    dialogues: List[Dict[str, Any]]

class EnvironmentData(BaseModel):
    """原始环境音数据格式"""
    scenes: List[Dict[str, Any]]
    effects: List[Dict[str, Any]]

class ConversionRequest(BaseModel):
    """格式转换请求"""
    dialogueData: Dict[str, Any] = Field(..., description="角色对话数据")
    environmentData: Optional[Dict[str, Any]] = Field(None, description="环境音数据")
    backgroundMusic: Optional[Dict[str, Any]] = Field(None, description="背景音乐配置")
    projectInfo: Optional[Dict[str, Any]] = Field(None, description="项目基本信息") 

File: frontend/src/api/multitrackProject.js
import axios from 'axios'

const API_BASE = '/api/v1/multitrack'

// 项目管理接口
export async function createProject(projectData) {
  const res = await axios.post(`${API_BASE}/create`, { project: projectData })
  return res.data
}

export async function loadProject(projectId) {
  const res = await axios.get(`${API_BASE}/load/${projectId}`)
  return res.data
}

export async function saveProject(projectId, projectData) {
  const res = await axios.put(`${API_BASE}/save/${projectId}`, { project: projectData })
  return res.data
}

export async function listProjects() {
  const res = await axios.get(`${API_BASE}/list`)
  return res.data
}

export async function deleteProject(projectId) {
  const res = await axios.delete(`${API_BASE}/delete/${projectId}`)
  return res.data
}

// 格式转换接口
export async function convertToStandardFormat(conversionData) {
  const res = await axios.post(`${API_BASE}/convert`, conversionData)
  return res.data
}

// 项目验证接口
export async function validateProject(projectId) {
  const res = await axios.get(`${API_BASE}/validate/${projectId}`)
  return res.data
}

// 导出相关接口
export async function exportProject(projectId) {
  const res = await axios.post(`${API_BASE}/export/${projectId}`)
  return res.data
}

export async function getExportStatus(exportTaskId) {
  const res = await axios.get(`${API_BASE}/export/status/${exportTaskId}`)
  return res.data
}

export async function downloadExportedAudio(exportTaskId) {
  const response = await axios.get(`${API_BASE}/export/download/${exportTaskId}`, {
    responseType: 'blob'
  })
  
  // 创建下载链接
  const url = window.URL.createObjectURL(new Blob([response.data]))
  const link = document.createElement('a')
  link.href = url
  link.setAttribute('download', `multitrack_export_${exportTaskId}.wav`)
  document.body.appendChild(link)
  link.click()
  link.remove()
  window.URL.revokeObjectURL(url)
}

// 预览播放相关接口
export async function generatePreviewAudio(projectId, startTime = 0, duration = null) {
  const params = { start_time: startTime }
  if (duration !== null) {
    params.duration = duration
  }
  
  const res = await axios.post(`${API_BASE}/preview/${projectId}`, null, { params })
  return res.data
}

export function getPreviewAudioUrl(fileId) {
  const url = `http://localhost:8000${API_BASE}/preview/download/${fileId}`
  console.log('构建预览音频URL:', url)
  return url
}

export async function deletePreviewFile(filename) {
  const res = await axios.delete(`/api/v1/audio-editor/preview/${filename}`)
  return res.data
}

// 工具函数：创建空项目模板
export function createEmptyProject(title = '新建项目') {
  return {
    project: {
      id: '',
      title,
      description: '',
      author: 'AI-Sound',
      totalDuration: 0,
      sampleRate: 44100,
      channels: 2,
      bitDepth: 16,
      exportFormat: 'wav',
      createdAt: null,
      version: '1.0'
    },
    tracks: [
      {
        id: 'track_dialogue',
        name: '角色对话',
        type: 'dialogue',
        volume: 1.0,
        muted: false,
        solo: false,
        color: '#3498db',
        order: 1,
        clips: []
      },
      {
        id: 'track_environment',
        name: '环境音效',
        type: 'environment',
        volume: 0.8,
        muted: false,
        solo: false,
        color: '#27ae60',
        order: 2,
        clips: []
      },
      {
        id: 'track_background',
        name: '背景音乐',
        type: 'background',
        volume: 0.5,
        muted: false,
        solo: false,
        color: '#e74c3c',
        order: 3,
        clips: []
      }
    ],
    markers: []
  }
}

// 工具函数：计算项目总时长
export function calculateProjectDuration(project) {
  let maxDuration = 0
  
  project.tracks.forEach(track => {
    track.clips.forEach(clip => {
      const endTime = clip.startTime + clip.duration
      maxDuration = Math.max(maxDuration, endTime)
    })
  })
  
  return maxDuration
}

// 工具函数：时间格式化
export function formatTime(seconds) {
  const minutes = Math.floor(seconds / 60)
  const remainingSeconds = Math.floor(seconds % 60)
  return `${minutes}:${remainingSeconds.toString().padStart(2, '0')}`
}

// 工具函数：生成唯一ID
export function generateId(prefix = 'item') {
  return `${prefix}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
} 